File: examples/example2.py
#! /usr/bin/env python
#
#  Name:    example1.py
#
#  Purpose: This example performs a simple step scan of a motor
#           by creating and then executing an MX scan record.
#
#  Author:  William Lavender
#
#-------------------------------------------------------------------------
#
#  Copyright 2001 Illinois Institute of Technology
#
#  See the file "LICENSE" for information on usage and redistribution
#  of this file, and for a DISCLAIMER OF ALL WARRANTIES.
#
#-------------------------------------------------------------------------
#

import sys

sys.stdout = sys.stderr

#  Define the location of the MX installation directory.

mxdir = "/opt/mx"

#  Define the location of the example MX database.

database_filename = "example.dat"

#  Construct the name of the Mp package directory.

###mp_package_dir = [ "/home/lavender/mxdev/0.56.0/mp/libMp" ]

mp_package_dir = [ mxdir + "lib/mp" ]

#  Add the Mp package directory to the module load path.

sys.path[:0] = mp_package_dir

#  Load the Mp package.

import Mp

#  Read in the MX database used by this example.

record_list = Mp.setup_database( database_filename )

#  Find the devices to be used by the scan.

energy_motor_record = record_list.get_record("energy")

i_zero_record = record_list.get_record("Io")

i_trans_record = record_list.get_record("It")

timer_record = record_list.get_record("timer1")

#  Setup the scan parameters for a copper K edge scan.

start = 8950.0		# in eV
step_size = 1.0		# in eV
num_steps = 101
measurement_time = 1.0	# in seconds
settling_time = 0.1     # in seconds

#  Construct a scan description for the requested scan.

fmt = "example2scan scan linear_scan motor_scan \"\" \"\" "

fmt = fmt + "1 1 1 energy 2 Io It 0 %g preset_time \"%g timer1\" "

fmt = fmt + "text example2scan.out gnuplot "

fmt = fmt + "log($f[0]/$f[1]) %g %g %d"

description = fmt % \
	( settling_time, measurement_time, start, step_size, num_steps )

print "Scan description = '%s'\n" % ( description )

#  Add a corresponding scan record to the MX database of the current process.

scan_record = record_list.create_record_from_description( description )

scan_record.finish_record_initialization()

#  Perform the scan.

scan_record.perform_scan()

#  Delete the scan record from the client's database.
#
#  Strictly speaking, this is not really necessary since the example
#  program is about to exit anyway.  It is shown primarily to give
#  an example of how scan records are normally deleted in long running
#  application programs.

scan_record.delete_record()

File: examples/example2.c
/*
 * Name:    example2.c
 *
 * Purpose: This example performs a simple step scan of a motor
 *          by creating and then executing an MX scan record.
 *
 * Author:  William Lavender
 *
 *-------------------------------------------------------------------------
 *
 * Copyright 2000-2001, 2006 Illinois Institute of Technology
 *
 * See the file "LICENSE" for information on usage and redistribution
 * of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 *
 */

#include <stdio.h>
#include <stdlib.h>

#include "mx_record.h"
#include "mx_scan.h"

#define DATABASE_FILENAME	"example.dat"

int
main( int argc, char *argv[] )
{
	MX_RECORD *record_list;

	MX_RECORD *scan_record;

	char description[250];

	double start, step_size, measurement_time, settling_time;
	int num_steps;

	mx_status_type status;

	/* Read in the MX database used by this example. */

	status = mx_setup_database( &record_list, DATABASE_FILENAME );

	if ( status.code != MXE_SUCCESS ) {
		fprintf( stderr, "Cannot setup the MX database '%s'.\n",
				DATABASE_FILENAME );
		exit(1);
	}

	/* Enable the automatic display of scan progress plots. */

	mx_set_plot_enable( record_list, TRUE );

	/* Setup the scan parameters for a copper K edge scan. */

	start = 8950;			/* in eV */
	step_size = 1.0;		/* in eV */
	num_steps = 101;
	measurement_time = 1.0;		/* in seconds */
	settling_time = 0.1;		/* in seconds */

	/* Construct a scan description for the requested scan.  The string
	 * concatenation feature of ANSI C allows us to split the format
	 * string for sprintf() across several lines.
	 */

	sprintf( description,
		"example2scan scan linear_scan motor_scan \"\" \"\" "

		"1 1 1 energy 2 Io It 0 %g preset_time \"%g timer1\" "

		"text example2scan.out gnuplot log($f[0]/$f[1]) %g %g %d",

		settling_time, measurement_time,

		start, step_size, num_steps );

	/* Display the scan description. */

	printf( "Scan description = '%s'\n\n", description );

	/* Add a corresponding scan record to the MX database of the current
	 * process.
	 */

	status = mx_create_record_from_description( record_list, description,
					&scan_record, 0 );

	if ( status.code != MXE_SUCCESS ) {
		fprintf(stderr, "Cannot create the scan record.\n");
		exit(1);
	}

	status = mx_finish_record_initialization( scan_record );

	if ( status.code != MXE_SUCCESS ) {
		fprintf(stderr,
			"Cannot finish initialization of the scan record.\n");
		exit(1);
	}

	/* Perform the scan. */

	status = mx_perform_scan( scan_record );

	if ( status.code != MXE_SUCCESS ) {
		fprintf(stderr, "The scan aborted abnormally.\n" );
		exit(1);
	}

	/* Delete the scan record from the client's database.
	 *
	 * Strictly speaking, this is not really necessary since the example
	 * program is about to exit anyway.  It is shown primarily to give
	 * an example of how scan records are normally deleted in long running
	 * application programs.
	 */

	status = mx_delete_record( scan_record );

	if ( status.code != MXE_SUCCESS ) {
		fprintf(stderr,
			"The attempt to delete the scan record failed.\n");
		exit(1);
	}

	exit(0);
}

